TWAP logic (#2168)
osmosis-labs/osmosis: x/gamm/twap/README.md

# TWAP (Time Weighted Average Price)

The TWAP package is responsible for being able to serve TWAPs for every AMM pool.

A time weighted average price is a function that takes a sequence of `(time, price)` pairs, and returns a price representing an 'average' over the entire time period. The method of averaging can vary from the classic arithmetic mean, (such as geometric mean, harmonic mean), however we currently only implement arithmetic mean.

## Arithmetic mean TWAP

> Using the arithmetic mean, the TWAP of a sequence `(t_i, p_i)`, from `t_0` to `t_n`, indexed by time in ascending order, is: $\frac{1}{t_n - t_0}\sum_{i=0}^{n-1} p_i (t_{i+1} - t_i)$. (Notice that the latest price `p_n` isn't used, as it has lasted for a time interval of `0` seconds in this range!)

Author: hrmm, yeah this is weird. This is definitely valid latex syntax :confused:

> To illustrate with an example, given the sequence: `(0s, $1), (2s, $5), (5s, $1)`, the arithmetic mean TWAP is: $\frac{\$1 * (2s - 0s) + \$5 * (5s - 3s)}{5s - 0s} = \frac{\$10}{5} = \$2$.

Reviewer: Also i think calculation is  lil off 
```
(1 * 2 + 5 * 2 ) / 5  = 12/5
```

Also if we follow the formula wouldn't it be 
```
(1 * 2) + 5 * (5 - 2) / 5 = 15/5 
```

> To illustrate with an example, given the sequence: `(0s, $1), (2s, $5), (5s, $1)`, the arithmetic mean TWAP is: $\frac{\$1 * (2s - 0s) + \$5 * (5s - 3s)}{5s - 0s} = \frac{\$10}{5} = \$2$.

Author: ah yeah woops, thanks. Fixed the numbers.

I'm generally confused why github latex renders differently than most latex renderers :/

> To illustrate with an example, given the sequence: `(0s, $1), (2s, $5), (5s, $1)`, the arithmetic mean TWAP is: $\frac{\$1 * (2s - 0s) + \$5 * (5s - 3s)}{5s - 0s} = \frac{\$10}{5} = \$2$.

Reviewer: all good 👍

## Computation via accumulators method

The prior example for how to compute the TWAP takes linear time in the number of time entries in a range, which is too inefficient. We require TWAP operations to have constant time complexity (in the number of records).

> This is achieved by using an accumulator. In the case of an arithmetic Twap, we can maintain an accumulator from `a_n`, representing the numerator of the Twap expression for the interval `t_0...t_n`, namely $a_n = \sum_{i=0}^{n-1} p_i (t_{i+1} - t_i)$. If we maintain such an accumulator for every pool, with `t_0 = pool_creation_time` to `t_n = current_block_time`, we can easily compute the TWAP for any interval. The twap for the time interval of price points `t_i` to `t_j` is then $twap = \frac{a_j - a_i}{t_j - t_i}$, which is constant time given the accumulator values.

Reviewer 2: For consistency:
```suggestion
This is achieved by using an accumulator. In the case of an arithmetic TWAP, we can maintain an accumulator from `a_n`, representing the numerator of the TWAP expression for the interval `t_0...t_n`, namely $a_n = \sum_{i=0}^{n-1} p_i (t_{i+1} - t_i)$. If we maintain such an accumulator for every pool, with `t_0 = pool_creation_time` to `t_n = current_block_time`, we can easily compute the TWAP for any interval. The TWAP for the time interval of price points `t_i` to `t_j` is then $twap = \frac{a_j - a_i}{t_j - t_i}$, which is constant time given the accumulator values.
```

> This is achieved by using an accumulator. In the case of an arithmetic Twap, we can maintain an accumulator from `a_n`, representing the numerator of the Twap expression for the interval `t_0...t_n`, namely $a_n = \sum_{i=0}^{n-1} p_i (t_{i+1} - t_i)$. If we maintain such an accumulator for every pool, with `t_0 = pool_creation_time` to `t_n = current_block_time`, we can easily compute the TWAP for any interval. The twap for the time interval of price points `t_i` to `t_j` is then $twap = \frac{a_j - a_i}{t_j - t_i}$, which is constant time given the accumulator values.

Reviewer 2: > Notice that the latest price `p_n` isn't used, as it has lasted for a time interval of `0` seconds in this range

> If we maintain such an accumulator for every pool, with `t_0 = pool_creation_time` to `t_n = current_block_time`, we can easily compute the TWAP for any interval

Would accumulator at `t_n` be equal to `t_{n-1}` given the notice?

Author: No, because the accumulator at `t_n` (`a_n`) contains the result of `p_{n-1}`, wheres the accumulator at `t_{n-1}`, (`a_{n-1}`) does not have the result of multiplying by `p_{n-1}`

- [ ] Using table driven unit tests to test all foreseen cases the module can be within
    - hook testing
        - All swaps correctly trigger twap record updates
        - Create pools cause records to be created
    - store
        - EndBlock triggers all relevant twaps to be saved correctly
        - Block commit wipes temporary stores
    - logic
        - Make tables of expected input / output cases for:
          - getMostRecentRecord
          - getInterpolatedRecord
          - updateRecord
          - computeArithmeticTwap
        - Test overflow handling in all relevant arithmetic
        - Complete testing code coverage (up to return err lines) for logic.go file
    - API
        - Unit tests for the public API, under foreseeable setup conditions
- [ ] Integration into the Osmosis simulator
    - The osmosis simulator, simulates building up complex state machine states, in random ways not seen before. We plan on, in a property check, maintaining expected TWAPs for short time ranges, and seeing that the keeper query will return the same value as what we get off of the raw price history for short history intervals.

>     - Not currently in scope for release blocking, but planned: Integration for gas tracking, to ensure gas of reads/writes does not grow with time.

Reviewer 2: ```suggestion
    - Not currently in scope for the release, but planned: Integration for gas tracking, to ensure gas of reads/writes does not grow with time.
```

- [ ] Mutation testing usage
    - integration of the TWAP module into go mutation testing: https://github.com/osmosis-labs/go-mutesting

>         - The success we've seen with the tokenfactory module, is it succeeds at surfacing behavior for untested behavior.

Reviewer 2: ```suggestion
        - The success we've seen with the `tokenfactory` module. It succeeds at surfacing behavior for untested logic.
```

>           e.g. if you delete a line, or change the direction of a conditional, does a test catch it.

Reviewer 2: ```suggestion
          e.g. if you delete a line, or change the direction of a conditional, mutation tests show if regular Go tests catch it.
```

>     - We expect to get this to a state, where after mutation testing is ran, the only items it mutates, that is not caught in a test, is: Deleting `return err`, or `panic` lines, in the situation where that error return or panic isn't reachable.

Reviewer 2: - While this is awesome, do folks need to know such detail about mutation testing?
- Suggest paraphrasing to make it clearer. A lot of context for one sentence

	poolId uint64,
	baseAssetDenom string, quoteAssetDenom string,
	startTime time.Time, endTime time.Time) (sdk.Dec, error) { ... }
```

There are convenience methods for `GetArithmeticTwapToNow` which sets `endTime = ctx.BlockTime()`, and has minor gas reduction.
For users who need TWAPs outside the 48 hours stored in the state machine, you can get the latest accumulation store record from `GetBeginBlockAccumulatorRecord`.

## Code layout

**api.go** is the main file you should look at as a user of this module.

**logic.go** is the main file you should look at for how the TWAP implementation works.

- types/* - Implement TwapRecord, GenesisState. Define AMM interface, and methods to format keys.
- api.go - Public API, that other users / modules can/should depend on
- hook_listener.go - Defines hooks & calls to logic.go, for triggering actions on 
- keeper.go - generic SDK boilerplate (defining a wrapper for store keys + params)
- logic.go - Implements all TWAP module 'logic'. (Arithmetic, defining what to get/set where, etc.)
- module.go - SDK AppModule interface implementation.
- store.go - Managing logic for getting and setting things to underlying stores

## Store layout

> We maintain twap accumulation records for every AMM pool on Osmosis. 

Reviewer 2: ```suggestion
We maintain TWAP accumulation records for every AMM pool on Osmosis. 
```

Because Osmosis supports multi-asset pools, a complicating factor is that we have to store a record for every asset pair in the pool.
For every pool, at a given point in time, we make one twap record entry per unique pair of denoms in the pool. If a pool has `k` denoms, the number of unique pairs is `k * (k - 1) / 2`.

> Each twap record stores [(source)](https://github.com/osmosis-labs/osmosis/tree/main/proto/osmosis/gamm/twap):

Reviewer 2: page not found for the link

Author: Is now!

* last spot price of base asset A in terms of quote asset B
* last spot price of base asset B in terms of quote asset A
* Accumulation value of base asset A in terms of quote asset B
* Accumulation value of base asset B in terms of quote asset A

All TWAP records are indexed in state by the time of write.

A new TWAP record is created in two situations:
* When a pool is created

> * In the `EndBlock`, if the block contains any potentially price changing event for the pool. (Swap, LP, Exit)

Reviewer 2: Are (Swap, LP, Exit) exhaustive? Suggest specifying if this is an example or an exhaustive list

When a pool is created, records are created with the current spot price of the pool.

> During EndBlock, new records are created, with:

Reviewer 2: ```suggestion
During `EndBlock`, new records are created, with:
```

> * The accumulator's updated based upon the most recent prior accumulator's stored last spot price

Reviewer 2: ```suggestion
* The accumulator's value is updated based upon the most recent prior accumulator's stored last spot price
```

> * The LastSpotPrice's equal to the EndBlock spot price.

Reviewer 2: I'm guessing this is what was intended:
```suggestion
* The `LastSpotPrice` value is equal to the EndBlock spot price.
```
